Feature - Allow ordered_logistic to take array of integers as input

andrjohns opened the pull request:
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Adds functionality and testing for the ordered logistic distribution to take an array of integers and vector of continuous location variables as inputs - assuming that all have the same cutpoints (i.e. N responses on a given item).

What was this:
```
  for(n in 1:N)
    for(j in 1:J)
      y[n,j] ~ ordered_logistic(lambda[n,j], cut[j]);
    }
  }
```
Can now be written as:
```
  for(j in 1:J)
    y[j] ~ ordered_logistic(lambda[j], cut[j]);
  }
```

#### Intended Effect:
Simplify specification of ordered logistic distributions.

#### How to Verify:
```
./runTests.py test/unit/math/prim/mat/prob/ordered_logistic_test.cpp test/unit/math/prim/mat/prob/ordered_logistic_log_test.cpp
```

#### Side Effects:
The Stan parser will need to be updated to add signature:
```
int[] ~ ordered_logistic(vector,vector)
```
#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):
Andrew Johnson
By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_log.hpp:
```diff
@@ -51,6 +52,52 @@ namespace stan {
       return ordered_logistic_lpmf<T_lambda, T_cut>(y, lambda, c);
     }
 
+    /**
```
The first argument (`int` or `std::vector<int>`) could just be templated out in this file.  It will only be able to be instantiated with `int` or `std::vector<int>` because the function to which it delegates has that restriction.  The doc will have to be generalized a bit.

This would be nice to do, but it's optional for this pull request.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_log.hpp:
```diff
@@ -51,6 +52,52 @@ namespace stan {
+     * @param c Positive increasing vector of cutpoints.
+     * @return Log probability of outcome given location and
+     * cutpoints.
+
```
Thanks for all the doc---you don't need spaces between the different `@` types.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -55,8 +58,13 @@ namespace stan {
-    ordered_logistic_lpmf(int y, const T_lambda& lambda,
-                         const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c) {
+    ordered_logistic_lpmf(int y,
+                      const T_lambda& lambda,
```
We'd rather keep as many arguments as possible on a line.

The running theme here is conserving vertical space so it's easier to scan a program.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -73,25 +81,101 @@ namespace stan {
+      check_finite(function, "Location parameter", lambda[i]);
       }
+      for (int i = 1; i < c.size(); ++i)
+        check_greater(function, "Cut points parameter", c(i), c(i - 1));
```
Is there no `check_ordered`?  If not, there should be, as this won't produce a good message because `i` will be lost.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -73,25 +81,101 @@ namespace stan {
+      check_consistent_sizes(function, "Integers", y, "Locations", lambda);
+
+      for (int i = 0; i < N; ++i) {
+      check_bounded(function, "Random variable", y[i], 1, K);
```
indent the bodies of loops

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -73,25 +81,101 @@ namespace stan {
+      check_finite(function, "Cut points parameter", c(c.size()-1));
+      check_finite(function, "Cut points parameter", c(0));
+
+      // log(1 - inv_logit(lambda))
```
delete this stray comment

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -73,25 +81,101 @@ namespace stan {
-      return log_inv_logit_diff(c(y-2) - lambda,
-                                c(y-1) - lambda);
+      check_greater(function, "Size of cut points parameter", c.size(), 0);
+      check_finite(function, "Cut points parameter", c(c.size()-1));
```
The first one can say "Final cut-point" and the second can say "First cut-point".

andrjohns commented:
Thanks for the quick review, much appreciated! I've addressed those comments. I'm still fairly new to c++, so I haven't done much with templating.

After some thought I've also added the option to specify an array of ordered cutpoints, so the signature would be:

```
int[] ~ (vector, vector[])
```

This would address the use case where the array of integers are an individual's responses to J different items. So instead of:
```
for(n in 1:N)
  for(j in 1:J)
    y[n,j] ~ ordered_logistic(lambda[n,j], cut[j]);
```
You could write:
```
for(n in 1:N)
  y[n] ~ ordered_logistic(lambda[n], cut);
```

Let me know if there's anything else I can do with this to make it more efficient.

bob-carpenter commented:
Thanks for updating.

For templates, I'd recommend Vandevoorde and Josuttis's template book as well as Alexandrescu's. They need updates (as do I) for C++11.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_log.hpp:
```diff
@@ -20,37 +21,128 @@ namespace stan {
+                         const std::vector<Eigen::Matrix<
+                                                T_cut, Eigen::Dynamic, 1>>& c) {
+      return ordered_logistic_lpmf<T_loc, T_cut>(y, lambda, c);
+    }
```
With the additional signatures, you really need to generalize this.

You only need two template functions.  All the work, including ruling out illegal cases, will be done by the functions you're delegating to in the body of these two functions:

```
    template <bool propto, typename T_y, typename T_loc, typename T_cut>
    typename boost::math::tools::promote_args<T_loc, T_cut>::type
    ordered_logistic_log(const T_y& y, const T_loc& lambda,
                         const T_cut& c) {
      return ordered_logistic_lpmf<propto>(y, lambda, c);
    }

    template <typename T_y, typename T_loc, typename T_cut>
    typename boost::math::tools::promote_args<T_loc, T_cut>::type
    ordered_logistic_log(const T_y& y, const T_loc& lambda,
                         const T_cut& c) {
      return ordered_logistic_lpmf(y, lambda, c);
    }
```

They should both be documented as deprecated.

I'm pretty sure you won't need the explicit template arguments in the calls in the body, but you may need to put them back if the type inference won't go through on its own.

You can generalize doc to talk about values or containers of values.

Then, you don't need to include any of the container files here (like `<vector>`).

andrjohns replied:
That template only worked for me when lambda was double, when calling it with lambda as a vector it returns the error
```
stan/lib/stan_math/stan/math/prim/mat/prob/ordered_logistic_log.hpp:15:14: error: 
      no viable conversion from returned value of type 'typename
      boost::math::tools::promote_args<double, double>::type' (aka 'double') to
      function return type 'typename
      boost::math::tools::promote_args<Matrix<double, -1, 1, 0, -1, 1>,
      vector<Matrix<double, -1, 1, 0, -1, 1>, allocator<Matrix<double, -1, 1, 0,
      -1, 1> > > >::type' (aka 'Eigen::Matrix<double, -1, 1, 0, -1, 1>')
      return ordered_logistic_lpmf<propto>(y, lambda, c);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Only worked for me (for both vectorised cases) with:
```
    template <bool propto, typename T_y, typename T_loc, typename T_cut>
    typename boost::math::tools::promote_args<T_loc, T_cut>::type
    ordered_logistic_log(const T_y& y, 
                         const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
                         const T_cut& c) {
      return ordered_logistic_lpmf<propto>(y, lambda, c);

    }

    template <typename T_y, typename T_loc, typename T_cut>
    typename boost::math::tools::promote_args<T_loc, T_cut>::type
    ordered_logistic_log(const T_y& y, 
                         const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
                         const T_cut& c) {
      return ordered_logistic_lpmf(y, lambda, c);
    }
```
But this then doesn't work when lambda is a double, needing the separate template

bob-carpenter replied:
Sorry about that---needs a more general metaprogram.  You're looking for `return_type`.  Anyway, no big deal if you don't want to fix this.  It's just messy to maintain with lots of implementations.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -67,31 +75,180 @@ namespace stan {
       check_greater(function, "Size of cut points parameter", c.size(), 0);
-      for (int i = 1; i < c.size(); ++i)
-        check_greater(function, "Cut points parameter", c(i), c(i - 1));
+      check_ordered(function, "Cut-points", c);
```
Great!  Glad this was there.  Some of our original code really needs to be updated. When we add new utility code, like `check_ordered`, we don't do a good job of finding all the places where it should be used to simplify old code.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -67,31 +75,180 @@ namespace stan {
+      check_finite(function, "First cut-point", c(0));
+
+      if (y == 1) {
+        logp_n += -log1p_exp(lambda - c[0]);
```
You can just use `-=` here and in other cases--- I didn't catch this the first time.  No big deal if you don't want to change it.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -67,31 +75,180 @@ namespace stan {
+      for (int i = 0; i < N; ++i) {
+        int K = c[i].size() + 1;
+
+        check_bounded(function, "Random variable", y[i], 1, K);
```
These should still be outside of the loop so that you can use vectorized version.  the vectorized ones give more sensible error messages because they include the index.  here, we don't want to put it in the message because it takes too long to do all the string fiddling---it has to be done inside the test.

andrjohns commented:
Looks like the second edition of Vandevoorde and Josuttis's book is out with updates for c++11, perfect. Thanks!

andrjohns commented:
The templates for ```ordered_logistic_log``` are now all sorted and working without issue.

There were a couple of the checks that I could pull out of the loop (checking that values were finite), but the rest couldn't be pulled out without compile errors (since each element of the array is a vector). Let me know if there's another way of vectorising those checks.

Unless there were more changes this should also be safe to test with Jenkins now as well.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -46,55 +51,196 @@ namespace stan {
+      check_finite(function, "Final cut-point", c(c.size()-1));
+      check_finite(function, "First cut-point", c(0));
+
+      if (y == 1) {
```
You could've used the vector view pattern on the final argument to vectorize.  Not sure if it'd work for the integer arguments.

For this particular case, you could've avoided the local `logp_n` and just returned the results directly.  It's more efficient than allocating and incrementing for autodiff types.

Anyway, we can clean this up later, so no need to fix for this request.

bob-carpenter commented on stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp:
```diff
@@ -46,55 +51,196 @@ namespace stan {
+    ordered_logistic_lpmf(const std::vector<int>& y,
+                          const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
+                          const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c) {
+      typename return_type<T_loc>::type logp_n(0.0);
```
As a matter of style, you should declare variables as far down as you can so that their first use is as close to the declaration as possible.  Here that'd mean right before the for loop.  This also saves doing an allocation in case you throw, but that's not an issue here since throwing will clean everything anyway.

Again, no need to change here, just a heads up for style in the future.

bob-carpenter commented:
Jenkins, retest this please.

bob-carpenter commented:
@seantalts  I was going to merge this, but it looks like an issue with Jenkins not running.  Any idea what's going on here?  I still don't really get how we're supposed to test external pull requests.

bob-carpenter commented:
@seantalts  never mind on that last comment (erased)---it did kick off Jenkins, it's just slower to respond than it used to be.

seantalts commented:
Jenkins, retest this please.

andrjohns commented:
@seantalts I think the jenkins build for this one is stuck, would you mind having a look?

seantalts commented:
Jenkins, retest this please.

seantalts commented:
Jenkins, ok to test.
Jenkins, retest this please.
